Firestore/core/src/firebase/firestore/remote/stream.h

```diff
@@ -0,0 +1,193 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_REMOTE_GRPC_STREAM_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_REMOTE_GRPC_STREAM_H_
+
+#if !defined(__OBJC__)
+#error "This header only supports Objective-C++"
+#endif  // !defined(__OBJC__)
+
+#include <grpc/grpc.h>
+#include <grpcpp/client_context.h>
+#include <grpcpp/completion_queue.h>
+#include <grpcpp/generic/generic_stub.h>
+#include <grpcpp/security/credentials.h>
+#include <grpcpp/support/byte_buffer.h>
+
+#include "Firestore/core/include/firebase/firestore/firestore_errors.h"
+#include "Firestore/core/src/firebase/firestore/auth/credentials_provider.h"
+#include "Firestore/core/src/firebase/firestore/auth/token.h"
+#include "Firestore/core/src/firebase/firestore/remote/datastore.h"
+#include "Firestore/core/src/firebase/firestore/remote/exponential_backoff.h"
+#include "Firestore/core/src/firebase/firestore/remote/grpc_operation.h"
+#include "Firestore/core/src/firebase/firestore/remote/grpc_stream.h"
+#include "Firestore/core/src/firebase/firestore/remote/grpc_stream_objc_bridge.h"
+#include "Firestore/core/src/firebase/firestore/util/async_queue.h"
+#include "Firestore/core/src/firebase/firestore/util/executor.h"
+#include "Firestore/core/src/firebase/firestore/util/status.h"
+#include "absl/strings/string_view.h"
+
+#include <memory>
+
+#import "Firestore/Protos/objc/google/firestore/v1beta1/Firestore.pbobjc.h"
+#import "Firestore/Source/Core/FSTTypes.h"
+#import "Firestore/Source/Remote/FSTSerializerBeta.h"
+#import "Firestore/Source/Remote/FSTStream.h"
+#import "Firestore/Source/Util/FSTDispatchQueue.h"
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+class Stream : public GrpcOperationsObserver, public std::enable_shared_from_this<Stream> {
+ public:
+  Stream(util::AsyncQueue* async_queue,
+         auth::CredentialsProvider* credentials_provider,
+         Datastore* datastore,
+         util::TimerId backoff_timer_id,
+         util::TimerId idle_timer_id);
+
+  void Start();
+  void Stop();
+  bool IsStarted() const;
+  bool IsOpen() const;
+
+  void OnStreamStart() override;
+  void OnStreamRead(const grpc::ByteBuffer& message) override;
+  void OnStreamWrite() override;
+  // TODO OBC set state to error immediately when operation failed, or only in
+  // OnServerError?
+  void OnStreamError(const util::Status& status) override;
+
+  // ClearError?
+  void CancelBackoff();
+  void MarkIdle();
+  void CancelIdleCheck();
+
+  int generation() const override {
+    return generation_;
+  }
+
+ protected:
+  void EnsureOnQueue() const;
+  void Write(grpc::ByteBuffer&& message);
+  void ResetBackoff();
+
+ private:
+  enum class State {
```

Comment:
Please (eventually) transfer the existing comments over here (copying from JS or Android for the most up-to-date ones)